repose-aggregator/core/repose-core-api/src/test/scala/org/openrepose/core/systemmodel/config/AndTest.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * _=_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=
+ * Repose
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Copyright (C) 2010 - 2015 Rackspace US, Inc.
+ * _-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_-_=_
+ */
+
+package org.openrepose.core.systemmodel.config
+
+import javax.servlet.ServletInputStream
+import org.junit.runner.RunWith
+import org.mockito.Matchers.any
+import org.mockito.Mockito.when
+import org.openrepose.commons.utils.servlet.http.HttpServletRequestWrapper
+import org.scalatest.junit.JUnitRunner
+import org.scalatest.mock.MockitoSugar
+import org.scalatest.{FunSpec, Matchers}
+import org.springframework.mock.web.MockHttpServletRequest
+
+@RunWith(classOf[JUnitRunner])
+class AndTest extends FunSpec with Matchers with MockitoSugar {
+  describe("evaluate") {
+    it("should return true when all of the sub criteria are true") {
+      val filterCriteriaOne = mock[FilterCriterion]
+      when(filterCriteriaOne.evaluate(any(classOf[HttpServletRequestWrapper]))).thenReturn(true)
+      val filterCriteriaTwo = mock[FilterCriterion]
+      when(filterCriteriaTwo.evaluate(any(classOf[HttpServletRequestWrapper]))).thenReturn(true)
+      val filterCriterion = new And()
+      val subCriteria = filterCriterion.getFilterCriteria
+      subCriteria.add(filterCriteriaOne)
+      subCriteria.add(filterCriteriaTwo)
+      val originalRequest = new MockHttpServletRequest
+      val inputStream = mock[ServletInputStream]
+      val httpServletRequestWrapper = new HttpServletRequestWrapper(originalRequest, inputStream)
+
+      filterCriterion.evaluate(httpServletRequestWrapper) shouldBe true
+    }
+
+    Seq(
+      (false, false, false),
+      (true, false, false),
+      (false, true, false),
+      (false, false, true),
+      (true, true, false),
+      (true, false, true),
+      (false, true, true)
+    ).foreach { case (retOne, retTwo, retToo) =>
```

Comment:
Man, we use combinatorics a bunch when generating tests, and it's really something that seems to be lacking in Java/Scala standard libraries.

So here we could use the Scala standard library like so:
```
(List.fill(3)(List(true, false))).flatten.combinations(3).map(_.permutations).flatten.toList
```

But that's ugly. Both in terms of readability, and also in terms of efficiency. I am tempted to write a small library to supplement Scala's combinatorics... but not today.

Reply:
The problem with that is there is one case that is added to the final list that can't exist for this test, `(true, true, true)`.
Of course we could filter that out, but I think that takes away from the table like simplicity of what is being tested.

Reply:
Ah. I didn't notice that here. Fair enough. This is all good.

Also, in this case, the permutation of combinations is still a mess, so I wouldn't use it anyway.